isis/src/mro/apps/hical/hical.cpp

```diff
@@ -0,0 +1,530 @@
+      if ( !SkipModule(hiprof) ) {
+        ZeroReverse zr(caldata, hiconf);
+        calVars->add(hiconf.getProfileName(), zr.ref());
+        ZrHist = zr.History();
+        if ( hiprof.exists("DumpModuleFile") ) {
+          zr.Dump(hiconf.getMatrixSource("DumpModuleFile",hiprof));
+        }
+      }
+      else {
+        calVars->add(hiconf.getProfileName(), HiVector(nsamps, 0.0));
+        ZrHist.add("Debug::SkipModule invoked!");
+      }
+
+/////////////////////////////////////////////////////////////////
+// ZeroDark removes dark current
+//
+      procStep = "ZeroDark module";
+      hiconf.selectProfile("ZeroDark");
+      hiprof =  hiconf.getMatrixProfile();
+      HiHistory ZdHist;
+      ZdHist.add("Profile["+ hiprof.Name()+"]");
+      if ( !SkipModule(hiprof) ) {
+        ZeroDark zd(hiconf);
+        calVars->add(hiconf.getProfileName(), zd.ref());
+        ZdHist = zd.History();
+        if ( hiprof.exists("DumpModuleFile") ) {
+          zd.Dump(hiconf.getMatrixSource("DumpModuleFile",hiprof));
+        }
+      }
+      else {
+        calVars->add(hiconf.getProfileName(), HiVector(nsamps, 0.0));
+        ZdHist.add("Debug::SkipModule invoked!");
+      }
+
+/////////////////////////////////////////////////////////////////
+// ZeroDarkRate removes dark current with a new approach compared
+// with ZeroDark.
+
+      procStep = "ZeroDarkRate module";
+      HiHistory ZdrHist;
+      ZdrHist.add("Profile[ZeroDarkRate]");
+      // Check for backwards compatibility with old config files
+      if (!hiconf.profileExists("ZeroDarkRate")) {
+        calVars->add("ZeroDarkRate", HiVector(nsamps, 0.0));
+        ZdrHist.add("Skipped, module not in config file");
+      }
+      else {
+        hiconf.selectProfile("ZeroDarkRate");
+        hiprof =  hiconf.getMatrixProfile();
+        if ( !SkipModule(hiprof) ) {
+          // Make sure we aren't applying ZeroDark and ZeroDarkRate
+          if ( !SkipModule(hiconf.getMatrixProfile("ZeroDark")) ){
+            QString mess = "You have enabled both the ZeroDark and the ZeroDarkRate modules."
+                           "This means you are attempting to remove the dark current twice with "
+                           "two different algorithms. This is not approved use of hical. "
+                           "Please disable one or the other module using the Debug::SkipModule "
+                           "in your configuration file.";
+            throw IException(IException::User, mess, _FILEINFO_);
+          }
+          ZeroDarkRate zdr(hiconf);
+          calVars->add(hiconf.getProfileName(), zdr.ref());
+          ZdrHist = zdr.History();
+          if ( hiprof.exists("DumpModuleFile") ) {
+            zdr.Dump(hiconf.getMatrixSource("DumpModuleFile",hiprof));
+          }
+        }
+        else {
+          calVars->add(hiconf.getProfileName(), HiVector(nsamps, 0.0));
+          ZdrHist.add("Debug::SkipModule invoked!");
+        }
+      }
```

Comment:
This block is the implementation of the new module. The outer if/else block is to account for working with old config files which may not have the ZeroDarkRate profile defined.